flight/PiOS/Common/pios_bmp280.c

```diff
@@ -0,0 +1,441 @@
+	dev->digP5 = (data[15] << 8) | data[14];
+	dev->digP6 = (data[17] << 8) | data[16];
+	dev->digP7 = (data[19] << 8) | data[18];
+	dev->digP8 = (data[21] << 8) | data[20];
+	dev->digP9 = (data[23] << 8) | data[22];
+
+	if (PIOS_BMP280_WriteCommand(BMP280_CONFIG, 0x00) != 0)
+		return -2;
+
+	dev->task = PIOS_Thread_Create(
+		PIOS_BMP280_Task, "pios_bmp280", BMP280_TASK_STACK, NULL, BMP280_TASK_PRIORITY);
+	if (dev->task == NULL)
+		return -3;
+
+	PIOS_SENSORS_Register(PIOS_SENSOR_BARO, dev->queue);
+
+	return 0;
+}
+
+
+/**
+ * Claim the BMP280 device semaphore.
+ * \return 0 if no error
+ * \return -1 if timeout before claiming semaphore
+ */
+static int32_t PIOS_BMP280_ClaimDevice(void)
+{
+	PIOS_Assert(PIOS_BMP280_Validate(dev) == 0);
+
+	return PIOS_Semaphore_Take(dev->busy, PIOS_SEMAPHORE_TIMEOUT_MAX) == true ? 0 : 1;
+}
+
+/**
+ * Release the BMP280 device semaphore.
+ * \return 0 if no error
+ */
+static int32_t PIOS_BMP280_ReleaseDevice(void)
+{
+	PIOS_Assert(PIOS_BMP280_Validate(dev) == 0);
+
+	return PIOS_Semaphore_Give(dev->busy) == true ? 0 : 1;
+}
+
+
+/**
+* Start the ADC conversion
+* \param[in] PRESSURE_CONV or TEMPERATURE_CONV to select which measurement to make
+* \return 0 for success, -1 for failure (to start conversion)
+*/
+static int32_t PIOS_BMP280_StartADC(void)
+{
+	if (PIOS_BMP280_Validate(dev) != 0)
+		return -1;
+
+	/* Start the conversion */
+	return(PIOS_BMP280_WriteCommand(BMP280_CTRL_MEAS, dev->oversampling));
+
+return 0;
+}
+
+/**
+ * @brief Return the delay for the current osr
+ */
+static int32_t PIOS_BMP280_GetDelay() {
+	if (PIOS_BMP280_Validate(dev) != 0)
+		return 100;
+
+	switch(dev->oversampling) {
+		case BMP280_STANDARD_RESOLUTION:
+			return 14;
+		case BMP280_HIGH_RESOLUTION:
+			return 23;
+		case BMP280_ULTRA_HIGH_RESOLUTION:
+			return 44;
+		default:
+			break;
+	}
+	return 44;
+}
+
+/**
+* Read the ADC conversion value (once ADC conversion has completed)
+* \return 0 if successfully read the ADC, -1 if failed
+*/
+static int32_t PIOS_BMP280_ReadADC(bool calcTemp)
+{
+	if (PIOS_BMP280_Validate(dev) != 0)
+		return -1;
+
+	uint8_t data[6];
+
+	/* Read and store results */
+
+	if (PIOS_BMP280_Read(BMP280_PRESS_MSB, data, 6) != 0)
+			return -1;
+
+	if (calcTemp) {
+		int32_t raw_temperature = (int32_t)((((uint32_t)(data[3])) << 12) | (((uint32_t)(data[4])) << 4) | ((uint32_t)data[5] >> 4));
+
+		varT1 =  ((((raw_temperature >> 3) - ((int32_t)dev->digT1 << 1))) * ((int32_t)dev->digT2)) >> 11;
+		varT2 = (((((raw_temperature >> 4) - ((int32_t)dev->digT1)) * ((raw_temperature >> 4) - ((int32_t)dev->digT1))) >> 12) * ((int32_t)dev->digT3)) >> 14;
+		T = varT1 + varT2;
+		dev->compensatedTemperature = (T * 5 + 128) >> 8;
+	}
+
+	int32_t raw_pressure = (int32_t)((((uint32_t)(data[0])) << 12) | (((uint32_t)(data[1])) << 4) | ((uint32_t)data[2] >> 4));
+
+	varP1 = ((int64_t)T) - 128000;
+	varP2 = varP1 * varP1 * (int64_t)dev->digP6;
+	varP2 = varP2 + ((varP1 * (int64_t)dev->digP5) << 17);
```

Comment:
I was looking at DS Revision 1.14, May 2015. But there is an ever newer version (1.15). IMO, it would make sense to use the code provided in the latest DS. They may have changed it for a reason.